Add Socket Manager API and Windows support

naritta opened the pull request:
This is continued from below.
https://github.com/fluent/serverengine/pull/25 Sharing sockets between multiprocess workers with Socket Manager
https://github.com/fluent/serverengine/pull/26 Windows support for signal handling and dealing with IO

frsyuki commented on lib/serverengine/process_manager.rb:
```diff
@@ -204,14 +219,31 @@ def tick(blocking_timeout=0)
 
-      ready_pipes, _, _ = IO.select(@rpipes.keys, nil, nil, blocking_timeout)
+      if ServerEngine.windows?
+        @rpipes.each{|r, m|
```
why is this section necessary?

naritta replied:
That's because Win32::Pipe::Server doesn't cause EOFError even when a child process exits.
(https://github.com/fluent/serverengine/pull/26/files#diff-1e5e735f236537f81cca69fb746da179R221)

Then I think we should adapt to child process exits by Process.waitpid2 and if child process exits, we have to close pipe.

naritta replied:
By the way, win32-pipe is not stable and hard to use, because there is many bugs.
Then I think we should use UDPSocket for Heartbeat if it is allowed.

frsyuki replied:
if my understanding is correct, pipe won't be invalid even when a connecting process exists because another client can connect there. That's the reason why EOFError doesn't happen.

As I commented on https://github.com/fluent/serverengine/pull/26#discussion_r44332439, another idea is to not implement heartbeat at all on windows. I think it's good enough and make it simple. it mans that `ProcessManager#tick` does nothing excepting `@monitors.delete_if {|m| !m.tick(time) }`. Any thoughts there?

naritta replied:
I got it. I think so, too.
I fixed it.

frsyuki commented on lib/serverengine/winsock.rb:
```diff
@@ -0,0 +1,193 @@
+    attach_function :accept, [:socket, :ptr, :int], :socket
+
+    rubydll_path = Dir.glob(RbConfig.expand("$(bindir)")+"/msvcrt-ruby*.dll").first
+    ffi_lib rubydll_path
```
is this necessary? I wonder that `attach_function :rb_w32_wrap_io_handle` works without ffi_lib because ruby.ddl is already linked to this process.

naritta replied:
I thought I can use, too. But in this ffi version, ruby.ddl is not linked.
Maybe, in old ffi version, It seems that we can call without ffi_lib.

frsyuki commented on lib/serverengine/winsock.rb:
```diff
@@ -0,0 +1,193 @@
+    attach_function :WSAAccept, [:socket, :ptr, :int, :int, :int], :socket
+    attach_function :accept, [:socket, :ptr, :int], :socket
+
+    rubydll_path = Dir.glob(RbConfig.expand("$(bindir)")+"/msvcrt-ruby*.dll").first
```
If my understanding is correct, there several versions of ruby for windows: msvc, cygwin, and mingw. I wonder this file name is always correct or not.

naritta replied:
This can adapt to ruby mswin32, mswin64 and mingw32. (msvcrt-ruby*.dll in mingw32 and mingw32, msvcr*-ruby*.dll in mswin64)
Cygwin is the different story here, because unix version run in cygwin.

I think it's enough because many users install ruby with Rubyinstaller or Visual Studio.
If it's wrong or there is better way, I'll fix it.

frsyuki commented:
@naritta Very good. Can I ask you following items to merge this awesome work?

* a) Could you add a simple test cases of SocketManager on Windows? (here: https://github.com/fluent/serverengine/pull/28/files#diff-96c336025ad565bd2282a761368b9dbcR22). You may assume that it will be an example usage of SocketManager on Windows.
* b) CI (Travis CI and AppVeyor) is failing. Could you check it?
* c) Could you squash changes into a simple commit or commits?
* d) It seems that win32-pipe is not used any more. Could you update gemspec file?
* e) If everything is ready to merge, please let me know.

naritta commented:
@frsyuki 
Thank you for the reviewing. I think it's ready to merge

I added AppVeyor support and I fixed previous problem in travis-ci.
https://github.com/fluent/serverengine/commit/b0a981d36787744f9841d003d95baa18b41359df#diff-180360612c6b8c4ed830919bbb4dd459R2
https://github.com/fluent/serverengine/commit/a5876394c9a832efb2d8ea40111b6f4e0539b9d5#diff-26195760b9c62e6865cd2a0136372463R25
There is other problem in Travis-ci, but I guess this is bugs in Travis-ci.
